homeassistant/components/humidifier/__init__.py

```diff
@@ -0,0 +1,308 @@
+"""Provides functionality to interact with humidifier devices."""
+from abc import abstractmethod
+from datetime import timedelta
+import logging
+from typing import Any, Dict, List, Optional
+
+import voluptuous as vol
+
+from homeassistant.const import SERVICE_TURN_OFF, SERVICE_TURN_ON
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.config_validation import (  # noqa: F401
+    ENTITY_SERVICE_SCHEMA,
+    PLATFORM_SCHEMA,
+    PLATFORM_SCHEMA_BASE,
+)
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.entity_component import EntityComponent
+from homeassistant.helpers.typing import ConfigType, HomeAssistantType
+
+from .const import (
+    ATTR_CURRENT_HUMIDITY,
+    ATTR_CURRENT_TEMPERATURE,
+    ATTR_FAN_MODE,
+    ATTR_FAN_MODES,
+    ATTR_HUMIDITY,
+    ATTR_HUMIDIFIER_ACTION,
+    ATTR_OPERATION_MODE,
+    ATTR_OPERATION_MODES,
+    ATTR_MAX_HUMIDITY,
+    ATTR_MIN_HUMIDITY,
+    ATTR_PRESET_MODE,
+    ATTR_PRESET_MODES,
+    ATTR_WATER_LEVEL,
+    DOMAIN,
+    OPERATION_MODE_HUMIDIFY,
+    OPERATION_MODE_DRY,
+    OPERATION_MODE_HUMIDIFY_DRY,
+    OPERATION_MODE_OFF,
+    OPERATION_MODES,
+    SERVICE_SET_FAN_MODE,
+    SERVICE_SET_HUMIDITY,
+    SERVICE_SET_OPERATION_MODE,
+    SERVICE_SET_PRESET_MODE,
+    SUPPORT_FAN_MODE,
+    SUPPORT_PRESET_MODE,
+    SUPPORT_TARGET_HUMIDITY,
+    SUPPORT_TEMPERATURE,
+    SUPPORT_WATER_LEVEL,
+    DEFAULT_MIN_HUMIDITY,
+    DEFAULT_MAX_HUMIDITY,
+)
+
+ENTITY_ID_FORMAT = DOMAIN + ".{}"
+SCAN_INTERVAL = timedelta(seconds=60)
+
+_LOGGER = logging.getLogger(__name__)
+
+SET_FAN_MODE_SCHEMA = ENTITY_SERVICE_SCHEMA.extend(
+    {vol.Required(ATTR_FAN_MODE): cv.string}
+)
+SET_PRESET_MODE_SCHEMA = ENTITY_SERVICE_SCHEMA.extend(
+    {vol.Required(ATTR_PRESET_MODE): cv.string}
+)
+SET_OPERATION_MODE_SCHEMA = ENTITY_SERVICE_SCHEMA.extend(
+    {vol.Required(ATTR_OPERATION_MODE): vol.In(OPERATION_MODES)}
+)
+SET_HUMIDITY_SCHEMA = ENTITY_SERVICE_SCHEMA.extend(
+    {vol.Required(ATTR_HUMIDITY): vol.Coerce(float)}
+)
+
+
+async def async_setup(hass: HomeAssistantType, config: ConfigType) -> bool:
+    """Set up humidifier devices."""
+    component = hass.data[DOMAIN] = EntityComponent(
+        _LOGGER, DOMAIN, hass, SCAN_INTERVAL
+    )
+    await component.async_setup(config)
+
+    component.async_register_entity_service(
+        SERVICE_TURN_ON, ENTITY_SERVICE_SCHEMA, "async_turn_on"
+    )
+    component.async_register_entity_service(
+        SERVICE_TURN_OFF, ENTITY_SERVICE_SCHEMA, "async_turn_off"
+    )
+    component.async_register_entity_service(
+        SERVICE_SET_FAN_MODE, SET_FAN_MODE_SCHEMA, "async_set_fan_mode"
+    )
+    component.async_register_entity_service(
+        SERVICE_SET_OPERATION_MODE,
+        SET_OPERATION_MODE_SCHEMA,
+        "async_set_operation_mode",
+    )
+    component.async_register_entity_service(
+        SERVICE_SET_PRESET_MODE, SET_PRESET_MODE_SCHEMA, "async_set_preset_mode"
```

Comment:
Pass the required supported feature to use this service.

Reply:
done